Packet++/src/TextBasedProtocol.cpp

```diff
@@ -9,685 +9,702 @@
-	range = m_FieldNameToFieldMap.equal_range(fieldName);
-	int i = 0;
-	for (std::multimap<std::string,HeaderField*>::iterator iter = range.first; iter != range.second; ++iter)
-	{
-		if (i == index)
+		// attach new field to message
+		newFieldToAdd->attachToTextBasedProtocolMessage(this, newFieldOffset);
+
+		// insert field into fields link list
+		if (prevField == nullptr)
 		{
-			fieldToRemove = iter->second;
-			break;
+			newFieldToAdd->setNextField(m_FieldList);
+			m_FieldList = newFieldToAdd;
+		}
+		else
+		{
+			newFieldToAdd->setNextField(prevField->getNextField());
+			prevField->setNextField(newFieldToAdd);
 		}
 
-		i++;
-	}
-
-	if (fieldToRemove != nullptr)
-		return removeField(fieldToRemove);
-	else
-	{
-		PCPP_LOG_ERROR("Cannot find field '" << fieldName << "'");
-		return false;
-	}
-}
+		// if newField is the last field, update m_LastField
+		if (newFieldToAdd->getNextField() == nullptr)
+			m_LastField = newFieldToAdd;
 
-bool TextBasedProtocolMessage::removeField(HeaderField* fieldToRemove)
-{
-	if (fieldToRemove == nullptr)
-		return true;
+		// insert the new field into name to field map
+		std::string fieldName = newFieldToAdd->getFieldName();
+		std::transform(fieldName.begin(), fieldName.end(), fieldName.begin(), ::tolower);
+		m_FieldNameToFieldMap.insert(std::pair<std::string, HeaderField*>(fieldName, newFieldToAdd));
 
-	if (fieldToRemove->m_TextBasedProtocolMessage != this)
-	{
-		PCPP_LOG_ERROR("Field isn't associated with this message");
-		return false;
+		return newFieldToAdd;
 	}
 
-	std::string fieldName = fieldToRemove->getFieldName();
-
-	// shorten layer and delete this field
-	if (!shortenLayer(fieldToRemove->m_NameOffsetInMessage, fieldToRemove->getFieldSize()))
+	bool TextBasedProtocolMessage::removeField(std::string fieldName, int index)
 	{
-		PCPP_LOG_ERROR("Cannot shorten layer");
-		return false;
-	}
+		std::transform(fieldName.begin(), fieldName.end(), fieldName.begin(), ::tolower);
 
-	// update offsets of all fields after this field
-	HeaderField* curField = fieldToRemove->getNextField();
-	shiftFieldsOffset(curField, 0-fieldToRemove->getFieldSize());
-//	while (curField != NULL)
-//	{
-//		curField->m_NameOffsetInMessage -= fieldToRemove->getFieldSize();
-//		if (curField->m_ValueOffsetInMessage != -1)
-//			curField->m_ValueOffsetInMessage -= fieldToRemove->getFieldSize();
-//
-//		curField = curField->getNextField();
-//	}
+		HeaderField* fieldToRemove = nullptr;
 
-	// update fields link list
-	if (fieldToRemove == m_FieldList)
-		m_FieldList = m_FieldList->getNextField();
-	else
-	{
-		curField = m_FieldList;
-		while (curField->getNextField() != fieldToRemove)
-			curField = curField->getNextField();
+		std::pair<std::multimap<std::string, HeaderField*>::iterator,
+		          std::multimap<std::string, HeaderField*>::iterator>
+		    range;
+		range = m_FieldNameToFieldMap.equal_range(fieldName);
```

Comment:
maybe we can apply the same fix we did in line 349 here?

Reply:
Fixed.